### PR TITLE
Speech Demo Sync

### PR DESCRIPTION
**Pull request by pluskid**
With some hard work from @yzhang87 , now we have a full pipeline including decoding with Kaldi to get the word error rates on a 100 hour AMI dataset, reproducing results closing to the state of the art. The LSTM training is implemented with the new module system and could be run in one of the two ways

- bucketing: limited to 800-frame unrolling but we break super-long sentences with overlapping in the IO.
- truncated-bptt (also implemented with the intermediate-level module API) with 20-frame unrolling.

### CODE REVIEW

**Comment by pluskid** [comment]
Should be able to implement the full LSTM typically used in speech once #2039 is merged.

**Comment by antinucleon** [comment]
Awesome. I will try new API and check on NLP task today or tomorrow.

**Comment by piiswrong** [comment]
what's the current state of the art on TIMIT?

**Comment by yzhang87** [comment]
@piiswrong 16%～·17% using bidirectional LSTM.

**Comment by tqchen** [comment]
@pluskid any luck trying the memonger to save memory cost of long LSTM?

**Comment by pluskid** [comment]
@tqchen, not tried yet. The bucketing up to 800 frame already works quite well. Due to exploding gradient problem we switched to truncated bptt with only 20 frame unroll. It takes quite small memory even without applying memonger--currently we are running at the same speed as cntk but using only half of the memory.

**Comment by yzhang87** [comment]
I think that (memonger) is more useful for the sequence level training
(e.g. bi-directional training).

On Tue, May 10, 2016 at 12:24 AM, Tianqi Chen <notifications@github.com>
wrote:

> @pluskid <https://github.com/pluskid> any luck trying the memonger to
> save memory cost of long LSTM?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/dmlc/mxnet/pull/2086#issuecomment-218055930>
>



-- 
Yu Zhang
Computer Science and Artificial Intelligence Laboratory (CSAIL)
Massachusetts Institute of Technology

**Comment by tqchen** [comment]
sounds great

**Comment by pluskid** [comment]
Yes, we are preferring bptt for gradient exploding reason. But there are more complicated scenarios that might not be easy to do with bptt as Yu pointed out

- bi-directional RNNs
- Sequence training (the loss function, unlike frame-wise softmax, is not easily decomposable according to the frames)
- end-to-end

So bucketing is needed for those complicated scenarios and memonger should prove to be useful in those cases.

**Comment by yzhang87** [comment]
@pluskid @tqchen after carefully scheduling the learning rate, gradient exploding might be OK for bucketing case although there are still performance gaps. 

But I think bi-directional RNNs would be a very good example to show the power of memonger. In current implementation we have to chop the utterance into smaller one (in Kaldi and CNTK) and add some contexts back (additional training cost but this is very important for the truncation case).
